fdbserver/team/TestBase.actor.cpp

```diff
@@ -0,0 +1,749 @@
+
+	// Debug ID
+	Optional<UID> debugID;
+
+	// Response
+	ReplyPromise<TLogCommitReply> reply;
+
+	TLogCommitRequest() {}
+	TLogCommitRequest(const SpanID& spanID,
+	                  const Arena arena,
+	                  StringRef messages,
+	                  const Version prevVersion,
+	                  const Version version,
+	                  const Version knownCommittedVersion,
+	                  const Version minKnownCommittedVersion,
+	                  const Optional<UID>& debugID)
+	  : spanID(spanID), arena(arena), messages(messages), prevVersion(prevVersion), version(version),
+	    knownCommittedVersion(knownCommittedVersion), minKnownCommittedVersion(minKnownCommittedVersion),
+	    debugID(debugID) {}
+
+	template <typename Ar>
+	void serialize(Ar& ar) {
+		serializer(ar,
+		           spanID,
+		           arena,
+		           messages,
+		           prevVersion,
+		           version,
+		           knownCommittedVersion,
+		           minKnownCommittedVersion,
+		           debugID,
+		           reply);
+	}
+};
+
+/**
+ * @class ProxyTLogPushMessageEncoder
+ * @brief Encode the mutations that commit proxy have received, for TLog's consumption
+ *
+ * The teamID is encoded first. Then the total number of the messages is recoreded.
+ *
+ * For each message, the format is:
+ *
+ *    | Size of the message | Subsequence     | Mutation      |
+ *    +---------------------+-----------------+---------------+
+ * 	  | uint32_t            | uint32_t        | variable bits |
+ *
+ * Finally, a CRC32 is calculated for all messages, including the leading teamID.
+ *
+ * And there is no gap between each message
+ */
+class ProxyTLogPushMessageEncoder {
+	/// Maps the TeamID to the list of BinaryWriters
+	std::unordered_map<TeamID, int> teamIDWriterMapping;
+	std::vector<BinaryWriter> writers;
+	std::vector<int> numMessages;
+
+	uint32_t currentSubsequence = 1;
+
+	/// Flag the encoding process is finished
+	bool finished = false;
+
+public:
+	void writeMessage(const MutationRef& mutation, const TeamID& teamID) {
+		ASSERT(!finished);
+
+		if (teamIDWriterMapping.find(teamID) == teamIDWriterMapping.end()) {
+			// An unseen TeamID
+			teamIDWriterMapping[teamID] = writers.size();
+			writers.emplace_back(BinaryWriter(AssumeVersion(g_network->protocolVersion())));
+			writers.back() << teamID << /* Number of messages */ uint32_t(0);
+			numMessages.push_back(0);
+		}
+		BinaryWriter& writer = writers[teamIDWriterMapping[teamID]];
+
+		const int lengthBeforeAppend = writer.getLength();
+		writer << /* Message size placeholder */ uint32_t(0) << (currentSubsequence++) << mutation;
+		const int lengthAfterAppend = writer.getLength();
+
+		*(reinterpret_cast<uint32_t*>(reinterpret_cast<uint8_t*>(writer.getData()) + lengthBeforeAppend)) =
+		    lengthAfterAppend - lengthBeforeAppend;
+
+		++numMessages[teamIDWriterMapping[teamID]];
+	}
+
+	Standalone<StringRef> getEncodedMessageForTeam(const TeamID& teamID) {
```

Comment:
Since you are serializing the data, it's very important to clearly define the format. I'd suggest define a message header with at least following fields: `message version`, `length`, `checksum`, and other information that are common for all the data in the message. When serializing the data, write the header and then the message body.